File: Backend/python-backend/src/utils/notification_helper.py
import requests
import os
from src.utils.logger import logger

JAVA_BACKEND_URL = os.getenv('JAVA_BACKEND_URL', 'http://localhost:5002')

class NotificationHelper:
    @staticmethod
    def send_notification(user_id, title, message, notification_type='INFO'):
        try:
            response = requests.post(
                f'{JAVA_BACKEND_URL}/system/notifications',
                json={
                    'userId': user_id,
                    'title': title,
                    'message': message,
                    'type': notification_type
                },
                timeout=5
            )
            response.raise_for_status()
            logger.info(f"Notification sent to user {user_id}: {title}")
        except Exception as e:
            logger.error(f"Failed to send notification: {e}")

    @staticmethod
    def notify_visitor_approval_request(host_user_id, visitor_name, purpose, visitor_id):
        NotificationHelper.send_notification(
            host_user_id,
            'Visitor Approval Required',
            f'{visitor_name} is waiting for approval. Purpose: {purpose}. Visitor ID: {visitor_id}',
            'WARNING'
        )

    @staticmethod
    def notify_incident_reported(user_id, severity, description):
        NotificationHelper.send_notification(
            user_id,
            'Incident Report Submitted',
            f'Your {severity} incident report has been submitted: {description[:50]}...',
            'WARNING'
        )

    @staticmethod
    def notify_safety_inspection(user_id, location, date):
        NotificationHelper.send_notification(
            user_id,
            'Safety Inspection Scheduled',
            f'Safety inspection scheduled at {location} on {date}.',
            'INFO'
        )

    @staticmethod
    def notify_safety_training(user_id, training_name, date):
        NotificationHelper.send_notification(
            user_id,
            'Safety Training Scheduled',
            f'You have been enrolled in {training_name} on {date}.',
            'INFO'
        )

notification_helper = NotificationHelper()

File: Backend/nodejs-backend/src/utils/notificationHelper.js
const axios = require('axios');

const JAVA_BACKEND_URL = process.env.JAVA_BACKEND_URL || 'http://localhost:5002';

class NotificationHelper {
  async sendNotification(userId, title, message, type = 'INFO') {
    try {
      await axios.post(`${JAVA_BACKEND_URL}/system/notifications`, {
        userId,
        title,
        message,
        type
      });
    } catch (error) {
      console.error('Failed to send notification:', error.message);
    }
  }

  async notifyLeaveRequestCreated(userId, leaveType, startDate, endDate) {
    await this.sendNotification(
      userId,
      'Leave Request Submitted',
      `Your ${leaveType} leave request from ${startDate} to ${endDate} has been submitted and is pending approval.`,
      'INFO'
    );
  }

  async notifyLeaveRequestApproved(userId, leaveType, startDate, endDate) {
    await this.sendNotification(
      userId,
      'Leave Request Approved',
      `Your ${leaveType} leave request from ${startDate} to ${endDate} has been approved.`,
      'SUCCESS'
    );
  }

  async notifyLeaveRequestRejected(userId, leaveType, startDate, endDate, reason) {
    await this.sendNotification(
      userId,
      'Leave Request Rejected',
      `Your ${leaveType} leave request from ${startDate} to ${endDate} has been rejected. Reason: ${reason || 'Not specified'}`,
      'WARNING'
    );
  }

  async notifyPayrollProcessed(userId, netPay, periodStart, periodEnd) {
    await this.sendNotification(
      userId,
      'Payroll Processed',
      `Your salary for ${periodStart} to ${periodEnd} has been processed. Net Pay: $${netPay.toFixed(2)}`,
      'SUCCESS'
    );
  }

  async notifyTrainingEnrollment(userId, courseName, startDate) {
    await this.sendNotification(
      userId,
      'Training Enrollment',
      `You have been enrolled in ${courseName} starting on ${startDate}.`,
      'INFO'
    );
  }

  async notifyPerformanceReview(userId, reviewDate) {
    await this.sendNotification(
      userId,
      'Performance Review Scheduled',
      `Your performance review has been scheduled for ${reviewDate}.`,
      'INFO'
    );
  }

  async notifyOnboarding(userId, firstName) {
    await this.sendNotification(
      userId,
      'Welcome to the Team!',
      `Welcome ${firstName}! Please complete your onboarding checklist and submit required documents.`,
      'INFO'
    );
  }

  async notifyOffboarding(userId, exitDate) {
    await this.sendNotification(
      userId,
      'Exit Process Initiated',
      `Your exit process has been initiated. Exit date: ${exitDate}. Please complete the clearance checklist.`,
      'WARNING'
    );
  }

  async notifyLeaveCompleted(userId, leaveType, startDate, endDate) {
    await this.sendNotification(
      userId,
      'Leave Period Completed',
      `Your ${leaveType} leave from ${startDate} to ${endDate} has ended. Welcome back!`,
      'INFO'
    );
  }
}

module.exports = new NotificationHelper();
